db/consistency_level.cc

```diff
@@ -23,8 +23,37 @@
 #include "locator/abstract_replication_strategy.hh"
 #include "locator/network_topology_strategy.hh"
 #include "heat_load_balance.hh"
+#include "tools/build_info.hh"
 
 namespace db {
+namespace {
+
+// Check the replicas data centers replication factor.
+// If we have nodes with RF=0, the effective replication map is inconsistent.
+void validate_replica_dc_replication_factor(const locator::effective_replication_map& erm, const inet_address_vector_replica_set& replicas) {
+    if constexpr (tools::build_info::is_release_build()) {
+        return;
+    }
+    const auto& topology = erm.get_topology();
+    const auto& replication_strategy = erm.get_replication_strategy();
+    // Applies only to Network Topology Replication Strategy.
+    if (replication_strategy.get_type() != locator::replication_strategy_type::network_topology) {
+        return;
+    }
+    inet_address_vector_replica_set nodes_with_errors;
+    nodes_with_errors.reserve(replicas.size());
+
+    const auto* network_topology_strategy = static_cast<const locator::network_topology_strategy*>(&replication_strategy);
+    std::ranges::copy_if(replicas, std::back_inserter(nodes_with_errors), [&network_topology_strategy, &topology](const auto& node) -> bool {
+        return network_topology_strategy->get_replication_factor(topology.get_datacenter(node)) == 0;
```

Comment:
Is it an error, when replicas set is has more nodes than block_for?

Reply:
My bad. It should be other way around. Not enough replicas to satisfy block_for().

Reply:
That check won't help in the case of our issue: we get block_for==0 and 1 node, so actually we get more replicas than block_for.

Reply:
our problem is that block_for is calculated basing on the "new picture" where the node is not counted, but replicas list still  reflects, the "old state" - replicas list contains the node as if it's data center's RF was not decreased to 0.

Reply:
0 is just a one case. Same may happen with decrease or increase to any value. With 0 we just happened to trigger a UB. My be the check should be that network_topology_strategy->get_replication_factor() is equal to erm->get_natural_endpoints(). In our case here the former will return 0 but the later will return 1.

Reply:
You mean to check if the number of nodes from a data center is equal to it's RF?

Reply:
The number of nodes returned as a replica for a token is the same as RF, yes. It may be smaller if number of nodes in a DC is smaller then configured RF though, but it never should be larger.

Reply:
Ok, then the check will be: 
For each data center the number of nodes should not be higher than the DC's replication factor.

Reply:
New version contains the check as agreed.